### PR TITLE
PHP 8 upgrade

### PR DESCRIPTION
**Pull request by jareddreyerss**
## PHP 8 upgrade
- Also includes new feature for client credential grant

### TODO:
- fix tests
- refactor controller so its not a weird mash up of a service and endpoints.

### CODE REVIEW

File: code/Entities/AccessTokenEntity.php [hunk 1]
```diff
@@ -65,7 +64,7 @@ public function getExpiryDateTime()
 
     public function getUserIdentifier()
     {
-        return $this->MemberID;
+        return $this->Client()->MemberID;
```
Comment (jareddreyerss) [inline code/Entities/AccessTokenEntity.php:70]: this isnt required, I abandoned the member object / login for oauth. But this wasn't connected to the client entity anyway so cool to leave it in the POC.

File: code/Entities/ClientEntity.php [hunk 2]
```diff
@@ -47,10 +54,11 @@ class ClientEntity extends DataObject implements ClientEntityInterface
         'ClientSecretHashMethod' => 'Varchar(50)',
         'ClientSecretHashIterations' => 'Varchar(50)',
         'ClientSecretSalt' => 'Varchar(50)',
+        'ClientConfidential' => 'Enum("true,false", "false")'
```
Comment (satrun77) [inline code/Entities/ClientEntity.php]: Use boolean.

File: code/Entities/ClientEntity.php [hunk 3]
```diff
@@ -163,6 +178,11 @@ public function getIdentifier()
         return $this->ClientIdentifier;
     }
 
+    public function setConfidential()
+    {
+        return $this->isConfidential = $this->ClientConfidential;
```
Comment (satrun77) [inline code/Entities/ClientEntity.php]: `isConfidential` does not exists or used in the dependency files

File: code/OAuthServerController.php [hunk 4]
```diff
@@ -83,11 +115,10 @@ public function __construct()
             $this->myRepositories['client'],
             $this->myRepositories['accessToken'],
             $this->myRepositories['scope'],
-            $privateKey,
-            $encryptionKey
+            $this->privateKey,
```
Comment (satrun77) [inline code/OAuthServerController.php:119]: `$this->server = @new AuthorizationServer(` the `@` should be removed, this is is part of the module core code, but it is a bad practise.

File: code/OAuthServerController.php [hunk 5]
```diff
@@ -240,11 +280,61 @@ public static function getMember($controller)
...
+     */
+    public function getServer()
+    {
+        return  $this->server;
+    }
+
+    /**
+     * @return ResourceServer|AuthorizationServer
+     */
+    public function setServer($server)
```
Comment (satrun77) [inline code/OAuthServerController.php:308]: Use parameter typehint

File: code/OAuthServerController.php [hunk 6]
```diff
@@ -240,11 +280,61 @@ public static function getMember($controller)
...
+    {
+        return $this->server = $server;
+    }
+
+    public function validateClientGrant(HTTPRequest $request)
+    {
+        $server = @new ResourceServer(
```
Comment (satrun77) [inline code/OAuthServerController.php]: Remove `@` and handle errors with exception

File: code/OAuthServerController.php [hunk 7]
```diff
@@ -240,11 +280,61 @@ public static function getMember($controller)
...
+            // Try to respond to the request
+            $this->myResponse = $this->server->validateAuthenticatedRequest($this->myRequest);
+
+        } catch (OAuthServerException $exception) {
+            // All instances of OAuthServerException can be formatted into a HTTP response
+            $this->myResponse = $exception->generateHttpResponse($this->myResponse);
+            return $this->myResponseAdapter->fromPsr7($this->myResponse);
+        } catch (Exception $exception) {
+            $this->myResponse = $this->myResponse->withStatus(500)->withBody(
+                stream_for($exception->getMessage())
```
Comment (satrun77) [inline code/OAuthServerController.php]: `stream_for` is deperated

File: code/OAuthServerController.php [hunk 8]
```diff
@@ -97,7 +128,7 @@ public function __construct()
         $grant->setRefreshTokenTTL(new DateInterval('P1M')); // refresh tokens will expire after 1 month
         $this->server->enableGrantType(
             $grant,
-            new DateInterval('PT1H') // access tokens will expire after 1 hour
+            new DateInterval($this->getGrantTypeExpiryInterval())
```
Comment (satrun77) [inline code/OAuthServerController.php]: `getGrantTypeExpiryInterval` is static method
```suggestion
            new DateInterval(self::getGrantTypeExpiryInterval())
```

File: code/Repositories/ClientRepository.php [hunk 9]
```diff
@@ -36,4 +38,20 @@ public function getClientEntity($clientIdentifier, $grantType = null, $clientSec
...
     }
+
+    /**
+     * @inheritDoc
+     */
+    public function validateClient($clientIdentifier, $clientSecret, $grantType)
+    {
+        $client = $this->getClientEntity($clientIdentifier);
+        
+        if ($client->ClientConfidential === true
```
Comment (satrun77) [inline code/Repositories/ClientRepository.php]:
```suggestion
        if ($client instanceof ClientEntity && $client->ClientConfidential === true
```

File: composer.json [hunk 10]
```diff
@@ -18,9 +18,10 @@
         "monolog/monolog": "^1.2",
```
Comment (satrun77) [inline composer.json:18]: Class `UserEntity` has deprecated method

**Review by satrun77 — CHANGES_REQUESTED** [review]

File: code/Entities/ClientEntity.php (hunk 3 above)
Comment (jareddreyerss) [reply code/Entities/ClientEntity.php]: yeah think we are ok, we don't need to store both and to not refactor the ss module I'm leaving it as ClientConfidential in the DB. isConfidential is member in league/oauth-server

File: code/OAuthServerController.php (hunk 8 above)
Comment (jareddreyerss) [reply code/OAuthServerController.php]: we're not using refreshtoken grant btw but yup can change this

File: composer.json (hunk 10 above)
Comment (jareddreyerss) [reply composer.json:18]: ignoring this one because we're not using `UserEntity` and its technically still working in php `7.4`

**Review by satrun77 — APPROVED** [review]

**Review by satrun77 — APPROVED** [review]

**Pull request merged**